packages/block-editor/src/store/private-actions.js

```diff
@@ -291,10 +296,30 @@ export function deleteStyleOverride( id ) {
 	};
 }
 
-export function syncDerivedBlockAttributes( clientId, attributes ) {
-	return {
-		type: 'SYNC_DERIVED_BLOCK_ATTRIBUTES',
-		clientIds: [ clientId ],
-		attributes,
+/**
+ * A higher-order action that mark every change inside a callback as "non-persistent"
+ * and ignore pushing to the undo history stack. It's primarily used for synchronized
+ * derived updates from the block editor without affecting the undo history.
+ *
+ * @param {() => void} callback The synchronous callback to derive updates.
+ */
+export function syncDerivedUpdates( callback ) {
+	return ( { dispatch, select, registry } ) => {
+		registry.batch( () => {
+			// Mark every change in the `callback` as non-persistent.
+			dispatch( {
+				type: 'SET_EXPLICIT_PERSISTENT',
+				isPersistentChange: false,
+			} );
+			callback();
+			dispatch( {
+				type: 'SET_EXPLICIT_PERSISTENT',
+				isPersistentChange: undefined,
+			} );
+
+			// Ignore pushing undo stack for the updated blocks.
+			const updatedBlocks = select.getBlocks();
+			undoIgnoreBlocks.add( updatedBlocks );
```

Comment:
It would be nice to be more explicit here and store only updated blocks instead of every block from the store. Probably something for the future and before API is stable.

Reply:
It's storing the updated blocks reference in a `WeakSet` so it shouldn't create extra memory nor leak it. In a way it's actually faster to not do any transformation here 😆. We can probably achieve the same thing with a flag in the store, like what `isPersistentChange` does, but the higher-order reducer makes it unnecessarily hard 😅. We can revisit this part in a future API design.